src/libsyntax/parse/parser.rs

```diff
@@ -2056,7 +2085,7 @@ impl<'a> Parser<'a> {
             self.bump();
             Ok(Ident::with_empty_ctxt(name))
         } else {
-            self.parse_ident()
+            self.parse_ident_common(false)
```

Comment:
I don't think we should fix this in `parse_ident`.
The condition is "doesn't start like a struct literal", not "`is_reserved_ident`".
https://github.com/rust-lang/rust/issues/15980 just happens to use `return` as something that isn't a struct literal, but the block can start with any other statement and the fix won't work.
I'd rather move the fix into parsing code for struct literals and used "not a non-reserved ident followed by colon" as a condition.

Reply:
Could you provide incorrect code that you'd expect to fail to be addressed by this PR? I'm confused, as currently (and after this PR) [other types of statements will have the expected behavior (continue and point out the lack of `=>`)](https://play.rust-lang.org/?gist=d90449b9ac7a3cff607e78dcc5735e4a&version=stable).

Reply:
Also, for the record, I'm slowly going through the parser looking for places where the span of an error is context dependent on the current state parser, in this case for example, the reason a non reserved ident was found is because we're parsing a struct, [in the case where](https://play.rust-lang.org/?gist=fbfa66a59623c292e331073182574ec4&version=nightly)

```rust
fn main() {
if 3 == {
    4
}
}
```

is expecting `{` is because it wants to parse the block following the `3 == { 4 }` `if` condition, etc.

Do you feel this is too heavy handed? The main advantage I see of this is 1) these are easy enough mistakes to make and 2) the diagnostic doesn't show you enough context as it is now to identify why the problem is happening in the first place.

Reply:
I thought we get a fatal error from `Struct { non_ident }` from which we cannot recover, but after rereading it turns out not to be the case.
And ability to not recover from a keyword in `parse_ident` seems usable in other contexts too.